interfaces/src/db.ts

```diff
@@ -79,11 +87,131 @@ export type Direction = 'ASC' | 'DESC';
 export const ASC: 'ASC' = 'ASC';
 export const DESC: 'DESC' = 'DESC';
 
-export type Order = [string[], Direction];
+export interface Order {
+  path: string[];
+  direction: Direction;
+}
 
 export interface Query {
   filter: Filter;
   limit?: number;
   skip?: number;
   orderBy?: ReadonlyArray<Order>;
 }
+
+// tslint:disable-next-line:no-empty-interface
+export interface ClientContext {}
+
+export interface ServerOnlyContext {
+  tags?: { [key: string]: string };
+  logLevel?: string;
+  logExtra?: { [key: string]: any };
+  sampleRate?: number;
+}
+
+export interface UpdateOptions {
+  /**
+   * Used to mark an update originated from a specific operation (useful for optimistic UI).
+   * This is meant to be UUID generated by the client.
+   */
+  readonly operationId?: string;
+}
+
+export interface Version {
+  major: number;
+  minor: number;
+}
+
+export interface Patch {
+  readonly version: Version;
+  /**
+   * Used to mark an update originated from a specific operation (useful for optimistic UI).
+   * @see UpdateOptions
+   */
+  readonly operationId?: string;
+  readonly ops: Operation[];
+}
+
+export type KeyedPatches = Array<[string, Patch[]]>;
+
+export interface VersionedObject {
+  version: Version;
+  value: Serializable;
+}
+
+interface Patches {
+  /**
+   * Stores changes made to the document, meant to be used internally by poll().
+   */
+  patches: Patch[];
+}
+
+export interface StoredDocument extends VersionedObject, Patches {
+  updatedAt: number;
+}
+
+/**
+ * A deleted document in the database, used for updating and for
+ * streaming.
+ */
+export interface Tombstone extends Patches {
+  updatedAt: number;
+  // (0, 0) if the document was never there in the first place.
+  version: Version;
+}
+
+export interface DB {
```

Comment:
I think all methods of this interface should include a table name parameter (we'll use default for now).

Reply:
Let's discuss F2F.

Reply:
We did :-)
I explained why I don't know where to put it (client context?  parameter??).  It depends on understanding the feature, which in turns depends on designing it. So not done for now.